cluster-app-values-schema/README.md

```diff
@@ -0,0 +1,283 @@
+# Basic requirements for cluster app values schema
+
+Cluster apps are apps which are used to configure and provision clusters based on Cluster API in the Giant Swarm platform. Like all helm charts, these apps are configured using so-called values YAML files, which are passed to the app as a ConfigMap for cluster creation. The allowed structure of the values YAML is defined by the app's values schema.
+
+This RFC defines basic requirements for all cluster apps provided by Giant Swarm.
+
+## Overview
+
+- [R1: JSON Schema dialect (draft 2019-09) must be specified](#r1)
+- [R2: Schema must explicitly cover all allowed properties](#r2)
+- [R3: Array item schema must be defined](#r3)
+- [R4: Properties must have a title](#r4)
+- [R5: Properties should have descriptions](#r5)
+- [R6: Properties should provide examples](#r6)
+- [R7: Constrain values as much as possible](#r7)
+- [R8: Required properties must be marked as such](#r8)
+- [R9: Avoid `anyOf` and `oneOf`](#r9)
+- [R10: Use `deprecated` to phase out properties](#r10)
+- [R11: Provide valid string values where possible](#r11)
+
+## Background
+
+TODO: Some more info regarding the rationale.
+
+- In-place documentation of properties
+- Better validation to catch configuration errors before they are applied
+- Generated entry forms
+- Generated information display
+- Enabling processes and communication about explicit contracts instead of implicit behaviour
+
+## Definitions
+
+- **Cluster app**: App (as defined by the Giant Swarm app platform) that allows for provisioning a cluster in a Giant Swarm installation.
+- **JSON Schema**: declarative language that allows to annotate and validate documents that are compatible with JSON.
+- The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://www.rfc-editor.org/rfc/rfc2119.html)
+
+## Requirements
+
+Requirements carry identifiers with prefix `R` and a unique number, for easy referencing.
+
+### R1: JSON Schema dialect (draft 2019-09) must be specified {#r1}
+
+Each cluster app's values schema file MUST specify the schema dialect (also called draft) used via the `$schema` keyword on the top level, with the draft URI as a value.
+
+The draft URI MUST be `https://json-schema.org/draft/2019-09/schema`.
```

Comment:
It is possible to reference external schemas with different draft versions. This could be problematic if we rely on a certain draft version in happa and in schemalint.
I suggest we restrict external schemas usage to those with the correct draft.

Reply:
It's indeed getting complicated if a referenced schema uses a different draft than the main one. Does the JSON schema spec mention anything about how to handle this?

Reply:
I tried to pinpoint this in the specification, but I didn't find it. However, the author of the JSON schema library we use said that [here](https://github.com/santhosh-tekuri/jsonschema/issues/103#issuecomment-1408736354) and I definitely trust him to know the specification.